motionplan/tpSpaceRRT.go

```diff
@@ -362,107 +539,285 @@ func (mp *tpSpaceRRTMotionPlanner) extendMap(
 			bestDist = cand.dist
 		}
 	}
-	treeNode := bestCand.treeNode
-	newNode := bestCand.newNode
+	treeNode := bestCand.treeNode // The node already in the tree to which we are parenting
+	newNode := bestCand.newNode   // The node we are adding because it was the best extending PTG
 
 	ptgNum := int(newNode.Q()[0].Value)
-	randK := uint(newNode.Q()[1].Value)
+	randAlpha := newNode.Q()[1].Value
+	randDist := newNode.Q()[2].Value
 
-	trajK := mp.tpFrame.PTGs()[ptgNum].Trajectory(randK)
+	trajK, err := mp.tpFrame.PTGs()[ptgNum].Trajectory(randAlpha, randDist)
+	if err != nil {
+		return nil, err
+	}
 
+	arcStartPose := treeNode.Pose()
+	if invert {
+		arcStartPose = spatialmath.Compose(arcStartPose, spatialmath.PoseInverse(trajK[len(trajK)-1].Pose))
+	}
 	lastDist := 0.
 	sinceLastNode := 0.
 
-	for _, trajPt := range trajK {
-		if ctx.Err() != nil {
-			return nil
-		}
-		if trajPt.Dist > newNode.Q()[2].Value {
-			// After we've passed goalD, no need to keep checking, just add to RRT tree
-			break
-		}
-		trajState := &State{Position: spatialmath.Compose(treeNode.Pose(), trajPt.Pose)}
-		sinceLastNode += (trajPt.Dist - lastDist)
-
-		// Optionally add sub-nodes along the way. Will make the final path a bit better
-		if mp.algOpts.addIntermediate && sinceLastNode > mp.algOpts.addNodeEvery {
-			// add the last node in trajectory
-			addedNode = &basicNode{
-				q:      referenceframe.FloatsToInputs([]float64{float64(ptgNum), float64(randK), trajPt.Dist}),
-				cost:   treeNode.Cost() + trajPt.Dist,
-				pose:   trajState.Position,
-				corner: false,
+	var trajState *ik.State
+	if mp.algOpts.addIntermediate {
+		for i := 0; i < len(trajK); i++ {
+			trajPt := trajK[i]
+			if invert {
+				trajPt = trajK[(len(trajK)-1)-i]
+			}
+			if ctx.Err() != nil {
+				return nil, ctx.Err()
 			}
-			rrt.maps.startMap[addedNode] = treeNode
-			sinceLastNode = 0.
+			trajState = &ik.State{Position: spatialmath.Compose(arcStartPose, trajPt.Pose)}
+			if mp.algOpts.pathdebug {
+				if !invert {
+					mp.logger.Debugf("$FWDTREE,%f,%f\n", trajState.Position.Point().X, trajState.Position.Point().Y)
+				} else {
+					mp.logger.Debugf("$REVTREE,%f,%f\n", trajState.Position.Point().X, trajState.Position.Point().Y)
+				}
+			}
+			sinceLastNode += (trajPt.Dist - lastDist)
+
+			// Optionally add sub-nodes along the way. Will make the final path a bit better
+			// Intermediate nodes currently disabled on the goal map because they do not invert nicely
+			if sinceLastNode > mp.algOpts.addNodeEvery && !invert {
+				// add the last node in trajectory
+				addedNode = &basicNode{
+					q:      referenceframe.FloatsToInputs([]float64{float64(ptgNum), randAlpha, trajPt.Dist}),
+					cost:   trajPt.Dist,
+					pose:   trajState.Position,
+					corner: false,
+				}
+				rrt[addedNode] = treeNode
+				sinceLastNode = 0.
+			}
+			lastDist = trajPt.Dist
+		}
+		if mp.algOpts.pathdebug {
```

Comment:
Do we want to be committing this into main?

Reply:
This has been SO darn useful and is the fourth time I've had to implement and then remove it. I think it's worth hiding behind an internal feature flag.